### PR TITLE
OpenCL auto-tune tweaks. Add backward GWS runs.

### PR DESCRIPTION
**Pull request by magnumripper**
#3620

### CODE REVIEW

File: src/opencl_common.c [hunk 1]
```diff
@@ -1819,14 +1821,13 @@ void opencl_find_best_lws(size_t group_size_limit, int sequential_id,
...
-	numloops = (int)(size_t)(200000000ULL / (endTime - startTime));
+	cl_ulong roundup = endTime - startTime - 1;
+	numloops = (int)(size_t)((200000000ULL + roundup) / (endTime - startTime));
 
 	clear_profiling_events();
 
 	if (numloops < 1)
 		numloops = 1;
-	else if (numloops > 5)
-		numloops = 5;
```
Comment (magnumripper) [inline src/opencl_common.c]: I dropped this, I'm not sure what that limitation buys us (we're talking 200ms here).

File: src/opencl_common.c [hunk 2]
```diff
@@ -2056,9 +2057,6 @@ void opencl_find_best_gws(int step, int max_duration,
 		raw_speed = (kpc / (run_time / 1E9)) * mask_int_cand.num_int_cand;
 		speed = rounds * raw_speed;
 
-		if (run_time < min_time)
-			min_time = run_time;
-
```
Comment (magnumripper) [inline src/opencl_common.c]: These were artifacts, never used

File: src/opencl_common.c (hunk 1 above)
Comment (claudioandre-br) [reply src/opencl_common.c]: It is acceptable. But, for very fast formats, from now on, you can run unnecessary tests (when 5 is enough).

File: src/opencl_common.c [hunk 3]
```diff
@@ -1499,7 +1500,8 @@ static void clear_profiling_events()
 	}
 }
 
-// Do the proper test using different global work sizes.
+// Do a test run with a specific global work size, return total duration
+// (or return zero for error or liomits exceeded)
```
Comment (claudioandre-br) [inline src/opencl_common.c]: ?

Comment (magnumripper) [reply src/opencl_common.c]: That old comment would perhaps be proper for `opencl_find_best_gws()`. However, this comment is for `gws_test()` which tests a single GWS as (now) described. `opencl_find_best_gws()` calls `gws_test()` for each GWS.

Comment (magnumripper) [reply src/opencl_common.c]: ...and liomits is my stiff upper lip pronunciation of limits 🤣

File: src/opencl_common.c (hunk 1 above)
Comment (magnumripper) [reply src/opencl_common.c]: For very fast formats, we just get more samples (and incidentally we increase the chance of winding up the GPU to full speed).

**Pull request merged**